File: python3/find_the_difference.py
class Solution:
    def findTheDifference(self, s: str, t: str) -> str:
        hash_s, hash_t = {}, {}

        for i in range(len(t)):
            if i != len(s):
                if s[i] not in hash_s:
                    hash_s[s[i]] = 1
                else:
                    hash_s[s[i]] += 1
            if t[i] not in hash_t:
                hash_t[t[i]] = 1
            else:
                hash_t[t[i]] += 1
                
        for k, v in hash_t.items():
            if k in hash_s:
                if abs(v - hash_s[k]) == 1:
                    return k
            else:
                return k
        

File: golang/find_the_difference.go
package main

// Space	: O(n)
// Time		: O(n)

func findTheDifference(s string, t string) byte {
	if len(t) > len(s) {
		s, t = t, s
	}

	var def byte
	_, n := len(s), len(t)
	smap := map[byte]int{}
	tmap := map[byte]int{}

	i := 0
	for i < n {
		smap[s[i]]++
		tmap[t[i]]++
		i++
	}

	smap[s[i]]++

	for k, v := range smap {
		if _, ok := tmap[k]; !ok {
			return k
		}
		if v != tmap[k] {
			return k
		}
	}

	return def
}
